### ISSUE TITLE
Need interface to operate on cache keys

### ISSUE THREAD

**Issue by cabo** [obgm/libcoap#470]
For processing related requests, it would be useful to have a datastructure that a server application can use to compare cache keys between these requests.  For example, responding to a blockwise request could be done very easily by caching the entire response body, and parceling out that body for requests that carry the same cache key (minus the Block 2 option).
There are several ways such a cache key could be represented.  A truncated hash of all non-NoCacheKey options (minus one option given as a parameter, here Block2) plus request  body would probably work best.  This hash probably needs to be cryptographic so the mechanism cannot be used for for an attack.
We may not need a full 32-byte SHA-256 though.

**Comment by mrdeep1** [comment]
Are you suggesting that this is only done server side?

If 2 different clients initiate the same request that hashes into the same value (using your current hash definition) with each client having a different view of the data, what happens then?

Separately, I was thinking of providing something like
````
/**
 * The data passed to coap_add_data_large_response() is free'd
 * once transmitted if set.
 */
#define COAP_LARGE_DATA_FLAGS_FREE 0x1

/**
 * Adds the initial part of @p data to the @p response pdu.  If blocks are
 * required, then the underlying CoAP layer will respond with the appropriate
 * subsequent block when requested rather than by the application layer.  If
 * the data fits with a single packet, the whole of the data will be sent.
 * Adds a ETAG option that is the hash of the entire data for client re-assembly.
 * Used by a GET request handler.
 *
 * As libcoap can be sending / receiving packets for other sessions while this
 * blocked transfer is taking place, it is the responsibility of the
 * application to not change the data until it has all been transferred.
 * The @p data can be internally freed off on completion of the transfer with
 * COAP_LARGE_DATA_FLAGS_FREE set in @p flags.
 * Two new events have been added so that the application can detect whether
 * the transfer has been successful or not.
 *
 * @param resource   The resource the data is associated with.
 * @param session    The coap session.
 * @param request    The requesting pdu.
 * @param response   The response pdu.
 * @param token      The token taken from the (original) requesting pdu.
 * @param media_type The format of the data.
 * @param maxage     The maxmimum life of the data. If @c -1, then there
 *                   is no maxage.
 * @param length     The total length of the data.
 * @param data       The entire data block to transmit.
 * @param flags      If COAP_LARGE_DATA_FLAGS_FREE set, then free off @p data
 *                   at completion of transmission.
 *
 */
void
coap_add_data_large_response(struct coap_resource_t *resource,
                             struct coap_session_t *session,
                             coap_pdu_t *request,
                             coap_pdu_t *response,
                             const coap_binary_t *token,
                             uint16_t media_type,
                             int maxage,
                             size_t length,
                             const uint8_t* data,
                             int flags);

````
where all the subsequent BLOCK2 reponses are handled by libcoap, not the application.

**Comment by obgm** [comment]
I was looking into some utility functions for caching. I think this would be proper way forward. A function such as `coap_add_data_large_response()` might be a mechanism to interface with this.

**Comment by cabo** [comment]
Right.  `coap_add_data_large_response()` would need some help to find out whether a Block2 request actually matches this response or not.  That's where we would use the cache key hash.  There would also need to be some input from applications on context, e.g., maybe the answer in a DTLS connection is different from that outside one even if the cache-key is the same.

**Comment by mrdeep1** [comment]
@cabo Please see #499 and see if this Cache Key / Cache Entry support is what you are looking for.

**Comment by obgm** [comment]
Closed as PR #499 now provides the requested functionality.
